### ISSUE TITLE
CSLA-Core-NS :  Dataportal.NewPILeaseAsync does not call DataPortal_Create()

### ISSUE THREAD

**Issue by jhuerta013** [MarimerLLC/cslaforum#594]
**Question(s)**
I am working on my first Core-NS project and I'm having issues calling the DataPortal_Create for my object.  I'm I missing something. I've worked on other projects with Full .Net and they work fine. Can someone tell me if I missing something? If I step through the code it will not hit the DataPortal_Create. I am able to call DataPortal_Fetch. 

Should the DataPortal_Fetch be written like this? "private async Task DataPortal_Fetch(int id)" 
with async Task

Thank you in advance. 

My solution is broken down like so:

1. Dal
2.DalSQL
3.Library
4.MVC - Core

Controller (stripdown:
      [HttpPost]
        public async Task<IActionResult> NewLease(NewLeaseModel lvm)
        {
              .....
                LeaseEdit le = await LeaseEdit.NewPILeaseAsync();
              .....
        }

Business Object:

        public async static System.Threading.Tasks.Task<LeaseEdit> NewPILeaseAsync()
        {
            return await DataPortal.CreateAsync<LeaseEdit>();
        }

        [RunLocal]
        protected override void DataPortal_Create()
        {
            base.DataPortal_Create();
        }

        private async Task DataPortal_Fetch(int id)
        {
            using (var ctx = PI.Dal.DalFactory.GetManager())
            {
                var dal = ctx.GetProvider<PI.Dal.ILeaseDal>();
                var data = dal.Fetch(id);
                using (BypassPropertyChecks)
                {
                    id = data.id;
                   .......
                }
         }

**Version and Platform**
CSLA version: 4.7.100
OS: Windows 
Platform:  ASP.NET Core

**Comment by ajj7060** [comment]
Yes; I believe if you are NOT actually going across a network boundry, if you do DataPortal.CreateAsync then you'll need to declare your create method as `private new Task DataPortal_Create`.

**Comment by rockfordlhotka** [comment]
Your example code doesn't use an `await` in the DP_Fetch method, so there's no reason to mark it `async` - that should be giving you a compiler warning actually.

@ajj7060 is right about `RunLocal` forcing DP_Create to run on (in your case) the web server vs an app server. That attribute does nothing if the data portal is configured to run the "server-side code" locally though (which is the default).

But either way, if your UI (controller) code calls `DataPortal.Create<foo>()` or `DataPortal.CreateAsync<foo>()` it should invoke the DP_Create method.

**Comment by ajj7060** [comment]
@rockfordlhotka Actually I found that if I had `[RunLocal] protected override void DataPortal_Create()` and then invoked `DataPortal.CreateAsync<T>()`, I got an error that the _Create had to be async.  The proxy configured was HttpProxy, I think it may have complained about non-async even though the create was a RunLocal.

**Comment by jhuerta013** [comment]
First off thank you for the responses. Can you provide me small but complete sample of how the     private "async Task DataPortal_Fetch(int id)" method should look like with await. I don't have the slightest idea on how to write one.

**Comment by rockfordlhotka** [comment]
@jhuerta013 your code isn't _wrong_, but the only reason to make the server-side method async is if that method contains an await statement. If you declare a method as async and don't use await inside the method, the C# compiler will give you a warning.

It is important to understand that the data portal has two different sides: client-side and server-side. Using async methods on one side doesn't impact whether you use them on the other side - the two are separate. I wrote a blog post about [async in the data portal](http://www.lhotka.net/weblog/CSLADataPortalChangesInVersion45.aspx).

So your server-side DP_Fetch method _can_ be async, but should only be async if you use an await to talk to the DAL. For example:

```c#
private async Task DataPortal_Fetch()
{
    var dal = DataAccess.GetProvider<mytype>();
    var data = await dal.GetDataAsync();
    // load domain object's properties from data object
}
```
You can do this regardless of whether the client-side code is async or not. The reason you make the server-side method async is because inside that method you need to use await.

**Comment by jhuerta013** [comment]
Hello Guys,
I found the reason why my create method was not being called. The AddObjectAuthorizationRules where blocking the call to the DataPortal_Create(). I created a None standard sample project and notice that I was getting "domain authorization errors" then it hit me that might be some type of permission issue. I commented the Rules and it worked. That explains why the fetch worked and not the create. Since I ran CodeSmith to generated all programs they were all templates with rules. What a relief. Back to the races!

Thank you so much.
